v4.0.0: modernized API

Froredion opened the pull request:
### Added

- `Zone.fromBox(cframe, size)` - Creates optimized box-shaped zones using GetPartBoundsInBox
- `Zone.fromSphere(position, radius)` - Creates optimized spherical zones using GetPartBoundsInRadius
- `Zone:setZoneShape(enumIdOrName)` - Allows manual control over spatial query method
- `ZoneShape` enum with options: Auto, Box, Sphere
- Modern OverlapParams management with automatic reuse for better performance
- Comprehensive documentation on spatial query API usage

### Changed

- **BREAKING**: Replaced deprecated `Region3` with modern `CFrame + Size` approach
  - `zone.region` is now split into `zone.regionCFrame` and `zone.regionSize`
  - `zone.exactRegion` is now `zone.exactRegionCFrame` and `zone.exactRegionSize`
- Updated all `FilterType.Whitelist/Blacklist` to modern `FilterType.Include/Exclude`
- Optimized OverlapParams reuse throughout the codebase
- Enhanced spatial query method selection based on zone geometry
- Improved README with modern API comparison and performance tips

Froredion committed:
v4.0.0

Froredion committed:
added place tester rbxlx for the maintainer to easily test

Froredion commented:
PS: I have an automated formatter, so it unnecessarily realigned many of the source files.

1ForeverHD commented:
Many thanks for the Pull Request @Froredion, it looks like a collection of fantastic improvements, and I'm excited to fully dive in and review once a few other projects are wrapped up in 2026

imezx commented:
so i tested it in live game but..

<img width="968" height="329" alt="image" src="https://github.com/user-attachments/assets/c183a96c-d94e-4820-a55d-372f06e7d652" />

its using `.fromBox` with medium accuracy.
im also added `table.clear(self)` and `setmetatable(self, nil)` when calling .destroy on zone.

Froredion commented:
When you destroy a zone with `table.clear(self)` and `setmetatable(self, nil)`, the zone object's properties (like zone.regionCFrame, zone.regionSize) become nil. However, the ZoneController's heartbeat connections might still be running and trying to access these properties before the zone is fully deregistered.

Is this what you mean? If so, I can see why it's happening.

Froredion committed:
(untested) solution: https://github.com/1ForeverHD/ZonePlus/pull/84#issuecomment-3419374487

revert if this did not fix the issue.

imezx commented:
yeah seems its fixed and i dont think you have to do this:

```lua
local occupantsDict = zonesAndOccupants[zone] or {}
local occupantsPresent = false
for k, v in pairs(occupantsDict) do
    occupantsPresent = true
    break
end
```
it should be:
```lua
local occupantsDict = zonesAndOccupants[zone] or {}
local occupantsPresent = next(occupantsDict) ~= nil
```

also, i dont think if you still have to use `pairs(tbl)` in 2025.

Froredion commented:
Got it, will do the edit soon.

AstaWasTaken commented:
This seems fully Ai written, assuming it’s entirely ai since its from superbullet

Froredion commented:
I am the creator of Superbullet.

Froredion committed:
fully optimized and tested: (see desc)

- added a tester for zone destroying
- improved code maintainability
- added selene.toml for people who wants to contribute and has selene installed

Froredion commented:
> yeah seems its fixed and i dont think you have to do this:
> 
> ```lua
> local occupantsDict = zonesAndOccupants[zone] or {}
> local occupantsPresent = false
> for k, v in pairs(occupantsDict) do
>     occupantsPresent = true
>     break
> end
> ```
> 
> it should be:
> 
> ```lua
> local occupantsDict = zonesAndOccupants[zone] or {}
> local occupantsPresent = next(occupantsDict) ~= nil
> ```
> 
> also, i dont think if you still have to use `pairs(tbl)` in 2025.

I've added the following:
- added a tester for zone destroying
- improved code maintainability (your comments)
- added selene.toml for people who wants to contribute and has selene installed

I can say that all are now properly tested.
